azure-storage-cpp-lite/src/blob/blob_client_wrapper.cpp

```diff
@@ -623,108 +630,101 @@ namespace microsoft_azure {
+                const auto originalEtag = firstChunk.response().etag;
+                const auto length = static_cast<unsigned long long>(firstChunk.response().totalSize);
 
-                            if(offset + range > current)
-                            {
-                                current = offset + range;
-                            }
-                            //std::cout << "current: " << current << std::endl;
-                            writeCV.notify_all();
-                        }
+                // Create and truncate the target file.
+                auto fd = open(destPath.c_str(), O_CREAT|O_WRONLY, 0770);
+                if (-1 == fd) {
+                    return;
+                }
+                if (-1 == ftruncate(fd, length)) {
+                    close(fd);
+                    return;
+                }
+
+                // Persist the first chunk.
+                auto r = pwrite(fd, os.str().c_str(), firstChunk.response().size, 0);
+                while (r != static_cast<ssize_t>(firstChunk.response().size) && r != -1) {
+                    r = pwrite(fd, os.str().c_str(), firstChunk.response().size, 0);
+                }
+                if (-1 == r) {
+                    close(fd);
+                    return;
+                }
+                close(fd);
 
+                // Download the rest.
+                const auto left = length - firstChunk.response().size;
+                const auto chunk_size = std::max(DOWNLOAD_CHUNK_SIZE, (left + downloaders - 1)/ downloaders);
+                std::vector<std::future<int>> task_list;
+                for(unsigned long long offset = firstChunk.response().size; offset < length; offset += chunk_size)
+                {
+                    const auto range = std::min(chunk_size, length - offset);
+                    auto single_download = std::async(std::launch::async, [originalEtag, offset, range, this, &destPath, &container, &blob](){
+                        // Note, keep std::ios_base::in to prevent truncating of the file.
+                        auto output = std::ofstream(destPath.c_str(), std::ios_base::out |  std::ios_base::in);
+                        output.seekp(offset);
+                        auto chunk = m_blobClient->get_chunk_to_stream_sync(container, blob, offset, range, output);
+                        output.close();
+                        if(!chunk.success())
                         {
-                            std::lock_guard<std::mutex> lock(mutex);
-                            ++parallel;
-                            //std::cout << "parallel done: " << parallel << std::endl;
-                            cv.notify_one();
+                           // Looks like the blob has been replaced by smaller one - ask user to retry.
+                           if (constants::code_request_range_not_satisfiable == chunk.error().code) {
+                              return EAGAIN;
```

Comment:
at the moment I think this will propagate up to FUSE.  Might be a good idea to check and retry in this case, either here or in fileapis.cpp

Reply:
no, it will be propagated up to the user giving something like "temporary ..."